### ISSUE TITLE
When used in Node-RED there are intermittent 'tablesorter is not a function' and 'ts.addWidget is not a function'

### ISSUE THREAD

**Issue by cuxcrider** [Mottie/tablesorter#1698]
I saw the issues reported here:  https://github.com/Mottie/tablesorter/issues/1489
and here:   https://github.com/Mottie/tablesorter/issues/1179

I tried to resolve the issue based on the feedback there.  jQuery is part of Node-RED so it seems likely there are duplicate copies being loaded.  I tried using the noconflict approach but this did not work.  

Here is my head:
```
<link rel="stylesheet" href="http://pinodered.local:1880/tablesorter/tablesorter-master/css/theme.default.css">
<script type="text/javascript" src="http://pinodered.local:1880/tablesorter/tablesorter-master/docs/js/jquery-latest.min.js"></script>
<script type="text/javascript" src="http://pinodered.local:1880/tablesorter/tablesorter-master/js/jquery.tablesorter.js"></script>
<script type="text/javascript" src="http://pinodered.local:1880/tablesorter/tablesorter-master/js/jquery.tablesorter.widgets.js"></script> 
```

And here is my code.  Here I tried using a setTimeout but there are still intermittent failures:
```
<style>
table, th , td  {
  border: 1px solid grey;
  border-collapse: collapse;
  padding: 5px;
}
table tr:nth-child(odd) {
  background-color: #f1f1f1;
}
table tr:nth-child(even) {
  background-color: #ffffff;
}

</style>
</head>

<table id="boards" class="tablesorter" >
    <thead>
    <tr>
        <th>Serial #</th>
        <th>Channel #</th>
        <th>Status</th>
        <th>Value</th>
        <th>Setpoint</th>
    </tr>
    </thead>
    <tbody>
    <tr ng-repeat-start="(serialNum, serialVal) in msg.boardMsg.serialNum"></tr>
        <tr ng-repeat="(channelNum, channelVal) in serialVal.channel">
            <td> {{serialNum}} </td>
            <td>{{channelNum}}</td>
            <td ng-style = "{color : channelVal.status === 'stopped' ? 'red' : 'green'}">{{channelVal.status == "stopped" ? "stopped" : "running"}}</td>  
            <td>{{channelVal.value}}</td>
            <td>{{channelVal.setpoint}}</td>
        </tr>
    <tr ng-repeat-end></tr>
    </tbody>
    
</table>



<script type="text/javascript">
//the timeout is necessary to allow the source to load, it usually loads in under 100ms, but not always, so 200 is better choice
setTimeout(function() {
    $('#boards').tablesorter();
},500);
</script>
```
It's quite possible I am not implementing the noconflict approach appropriately (I am definitely new to all the html, jquery, css stuff).  So if you think that might fix this and have some code specific to my case here I would appreciate it.  Thanks!

**Comment by Mottie** [comment]
Hi @cuxcrider!

Try *not* loading jQuery at all... I mean remove this line:

```html
<script type="text/javascript" src="http://pinodered.local:1880/tablesorter/tablesorter-master/docs/js/jquery-latest.min.js"></script>
```

Also, instead of using a `setTimeout` to wrap your script, use a [document ready function](https://api.jquery.com/ready/#ready-handler):

```html
<script>
// the timeout is necessary to allow the source to load, it usually loads in under 100ms, but not always, so 200 is better choice
$(function() {
    $('#boards').tablesorter();
});
</script>
```

**Comment by cuxcrider** [comment]
@Mottie thanks for the quick reply.  So it looks like the src is the biggest offender.

I removed:  ```<script type="text/javascript" src="http://pinodered.local:1880/tablesorter/tablesorter-master/docs/js/jquery-latest.min.js"></script>```

The page loads without any errors now about 90% of the time.  I will still occassionally get a "tablesorter is not a function" but everything looks good and works fine regardless.  

![full](https://user-images.githubusercontent.com/8380140/67644463-36152500-f8e7-11e9-9d98-a7898924a57b.JPG)

The only problem has been occassionally the table is not fully expanding in height and has just a sliver of it visible with scroll bars to the right.

![sliver](https://user-images.githubusercontent.com/8380140/67644464-3ca39c80-f8e7-11e9-8d11-632035ccfeae.JPG)


I tried making my script look like this:
```
<script type="text/javascript">
$(document).ready(function() {
     $('#boards').tablesorter();
});
</script>
```
But I still get the tablesorter error occasionally.  Additionally, with the .ready function it will sometimes load and look perfect, but the table will not actually sort.

Does any of that info help explain what might be going on?

**Comment by cuxcrider** [comment]
Hi @Mottie,

I'm really new to all this stuff, so I just want to ask a simple question that will maybe help solve my problems.

If I call:  ```src="http://pinodered.local:1880/tablesorter/tablesorter-master/docs/js/jquery-latest.min.js"``` Is the ```jquery-latest.min.js``` literally jquery, or is it something else?  I would assume it's something different because it is in the tablesorter directory.

**Comment by Mottie** [comment]
Yes, that is the latest version of jQuery. But, from looking at your URL, it seems you're linking to a local server. I would recommend linking to a CDN version of jQuery (if you need it) and tablesorter.

Copy the link you need from here:
- jQuery: https://cdnjs.com/libraries/jquery
- tablesorter: https://cdnjs.com/libraries/jquery.tablesorter

**Comment by TheSin-** [comment]
that is a local Mac or avahi server he might be developing.  If that is a live site then you are right that won't work but if it's local and he is a dev it's fine.

**Comment by cuxcrider** [comment]
@Mottie and @TheSin- I'm not sure on all the terminology, so I'll just say what I'm doing:  
I'm just making a table using Node-RED deployed on a raspberry pi.  About 10 people on my local network will access the table just to view some data from devices running in our office. 

As I understand when I launch Node-RED, jquery is included, so I should not need jquery.

I tried the cdn source of tablesorter but I got the same results with errors, etc.  I opted to install tablesorter locally because I figured it would be faster.

What is the difference between `jquery.tablesorter.min.js` and `jquery.tablesorter.js`?  Does `.min` eliminate duplicate code that might be causing me trouble?

**Comment by TheSin-** [comment]
th e.local will work on the LAN for machines that have avahi or bonjour, things like windows won't be able able use it though unless you install iTunes which comes with bonjour.

**Comment by TheSin-** [comment]
the .min.js is just a minified version, it's not as big cause things like comments and white spaces are removed, but in essence it's the same file just one of more human readable than the other.

**Comment by cuxcrider** [comment]
I believe there are issues with jquery already being loaded by NodeRed and I cannot seem to get around that.  Fortunately Node-Red now has a fully custom UIBuilder, which will only load what you want it to load.  You have to embark on a fully custom journey at that point, but it works...

**Comment by Mottie** [comment]
If jQuery is already loaded, don't load it again. Use that loaded version to initialize tablesorter.

Follow the example provided here: https://github.com/TotallyInformation/node-red-contrib-uibuilder/wiki/Example:-JQuery#example-front-end-files

**Comment by cuxcrider** [comment]
yes absolutely.  UIBuilder works fine.  Node-REDs  semi-custom UI on the other hand makes you load your own jquery if you want to write custom scripts while also loading jquery when you deploy Node-RED.  I tired using the noconflict approach but that failed.  So the UIBuilder is the way to go.  So in summary I do not think there is a problem with Tablesorter.
